File: Microservices/IoT.House.Automation.Microservices.Arduino/IoT.House.Automation.Microservices.Arduino.Infra/RabbitMQ/Subscription/SubscriptionManager.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using IoT.House.Automation.Microservices.Arduino.Application.Interfaces;
using IoT.House.Automation.Microservices.Arduino.Application.MessageBroker.Events;

namespace IoT.House.Automation.Microservices.Arduino.Infra.RabbitMQ.Subscription
{
    public class SubscriptionManager
    {
        private readonly IDictionary<string, IList<Subscription>> _handlers
            = new Dictionary<string, IList<Subscription>>();

        public bool IsEmpty => !_handlers.Keys.Any();

        public IEnumerable<string> EventNames => _handlers.Keys;

        public event EventHandler<string> OnEventRemoved;
        public event EventHandler<string> OnEventAdded;

        public void AddSubscription<TEvent, TEventHandler>()
            where TEvent : Event
            where TEventHandler : IEventHandler<TEvent>
        {
            AddSubscription(
                typeof(TEventHandler),
                typeof(TEvent).Name,
                typeof(TEvent)
                );
        }
        
        public bool HasSubscriptionsForEvent(string eventName) => _handlers.ContainsKey(eventName);

        public IEnumerable<Subscription> GetHandlersForEvent<TEvent>()
            where TEvent : Event
        {
            var key = typeof(TEvent).Name;
            return GetHandlersForEvent(key);
        }

        public IEnumerable<Subscription> GetHandlersForEvent(string eventName) => _handlers[eventName];

        public Type GetEventTypeByName(string eventName) => _handlers[eventName]?.FirstOrDefault()?.EventType;

        public void RemoveSubscription<TEvent, TEventHandler>()
            where TEvent : Event
            where TEventHandler : IEventHandler<TEvent>
        {

            var eventName = typeof(TEvent).Name;
            var handlerToRemove = FindSubscriptionToRemove(eventName, typeof(TEventHandler));
            RemoveSubscription(eventName, handlerToRemove);
        }

        private void AddSubscription(Type handlerType, string eventName, Type eventType)
        {
            if (!HasSubscriptionsForEvent(eventName))
            {
                _handlers.Add(eventName, new List<Subscription>());
                OnEventAdded?.Invoke(this, eventName);
            }

            if (_handlers[eventName].Any(s => s.HandlerType == handlerType))
            {
                throw new ArgumentException(
                    $"Handler Type {handlerType.Name} already registered for '{eventName}'", nameof(handlerType));
            }

            _handlers[eventName].Add(Subscription.New(
                handlerType, eventType)
                );
        }

        private Subscription FindSubscriptionToRemove(string eventName, Type eventHandler)
        {
            return !HasSubscriptionsForEvent(eventName) 
                ? null 
                : _handlers[eventName].SingleOrDefault(s => s.HandlerType == eventHandler);
        }

        private void RemoveSubscription(
            string eventName,
            Subscription subsToRemove
            )
        {
            if (subsToRemove == null) return;
            _handlers[eventName].Remove(subsToRemove);

            if (_handlers[eventName].Any()) return;

            _handlers.Remove(eventName);
            OnEventRemoved?.Invoke(this, eventName);
        }
    }
}

File: Microservices/IoT.House.Automation.Microservices.Mobile/infrastructure/src/main/java/iot/house/automation/microservices/mobile/infra/rabbitMQ/subscription/SubscriptionManager.java
package iot.house.automation.microservices.mobile.infra.rabbitMQ.subscription;

import com.sun.jdi.request.DuplicateRequestException;
import iot.house.automation.microservices.mobile.application.interfaces.EventHandler;
import iot.house.automation.microservices.mobile.application.messaging.events.BaseEvent;
import lombok.Setter;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;

public class SubscriptionManager {
    private final Map<String, List<Subscription>> handlers;

    @Setter
    private Consumer<String> onEventRemoved;
    @Setter
    private Consumer<String> onEventAdded;

    public SubscriptionManager() {
        handlers = new HashMap<>();
    }

    public boolean isEmpty() {
        return handlers.isEmpty();
    }

    public boolean hasSubscriptionForEvent(String eventName) {
        return handlers.containsKey(eventName);
    }

    public List<Subscription> getHandlersForEvent(String eventName) {
        return handlers.get(eventName);
    }

    public <TEvent extends BaseEvent, TEventHandler extends EventHandler> void addSubscription(Class<TEvent> eventType, Class<TEventHandler> handlerType) {
        addSubscription(eventType, eventType.getSimpleName(), handlerType);
    }

    public <TEvent extends BaseEvent, TEventHandler extends EventHandler> void removeSubscription(Class<TEvent> eventType, Class<TEventHandler> handlerType) {
        var eventName = eventType.getSimpleName();
        var handlerToRemove = findSubscriptionToRemove(eventName, handlerType);
        removeSubscription(eventName, handlerToRemove);
    }

    private <TEventHandler extends EventHandler> Subscription findSubscriptionToRemove(String eventName, Class<TEventHandler> handlerType) {
        return !hasSubscriptionForEvent(eventName)
                ? null
                : handlers.get(eventName).stream().filter(s -> s.getHandlerType() == handlerType).findFirst().orElse(null);
    }

    private <TEvent extends BaseEvent, TEventHandler extends EventHandler> void addSubscription(Class<TEvent> eventType, String eventName, Class<TEventHandler> handlerType) {
        if(!hasSubscriptionForEvent(eventName)){
            handlers.put(eventName, new ArrayList<>());
        }

        var handler = handlers.get(eventName);

        if(handler.stream().anyMatch(s -> s.getHandlerType() == handlerType)){
            throw new DuplicateRequestException(String.format("Handler Type %s already registered for '%s'", handlerType.getSimpleName(), eventName));
        }

        handler.add(Subscription.New(handlerType, eventType));
        onEventAdded.accept(eventName);
    }

    private void removeSubscription(String eventName, Subscription subscription) {
        if(subscription == null) return;
        handlers.get(eventName).remove(subscription);

        if(handlers.get(eventName).size() > 0) return;

        handlers.remove(eventName);
        onEventRemoved.accept(eventName);
    }
}
